src/protocols/gamma_control.rs

```diff
@@ -0,0 +1,237 @@
+    previous_gamma_ramp: Option<Vec<u16>>,
+    output: WlOutput,
+}
+
+impl GammaControlManagerState {
+    pub fn new<D>(display: &DisplayHandle) -> Self
+    where
+        D: GlobalDispatch<ZwlrGammaControlManagerV1, ()>,
+        D: Dispatch<ZwlrGammaControlManagerV1, ()>,
+        D: Dispatch<ZwlrGammaControlV1, GammaControlState>,
+        D: GammaControlHandler,
+        D: 'static,
+    {
+        display.create_global::<D, ZwlrGammaControlManagerV1, _>(VERSION, ());
+
+        Self {
+            gamma_controls: HashMap::new(),
+        }
+    }
+    pub fn destroy_gamma_control(&mut self, output_id: ObjectId) {
+        self.gamma_controls.remove(&output_id);
+    }
+}
+
+impl<D> GlobalDispatch<ZwlrGammaControlManagerV1, (), D> for GammaControlManagerState
+where
+    D: GlobalDispatch<ZwlrGammaControlManagerV1, ()>,
+    D: Dispatch<ZwlrGammaControlManagerV1, ()>,
+    D: Dispatch<ZwlrGammaControlV1, GammaControlState>,
+    D: GammaControlHandler,
+    D: 'static,
+{
+    fn bind(
+        _state: &mut D,
+        _handle: &DisplayHandle,
+        _client: &Client,
+        manager: New<ZwlrGammaControlManagerV1>,
+        _manager_state: &(),
+        data_init: &mut DataInit<'_, D>,
+    ) {
+        data_init.init(manager, ());
+    }
+}
+
+impl<D> Dispatch<ZwlrGammaControlManagerV1, (), D> for GammaControlManagerState
+where
+    D: GlobalDispatch<ZwlrGammaControlManagerV1, ()>,
+    D: Dispatch<ZwlrGammaControlManagerV1, ()>,
+    D: Dispatch<ZwlrGammaControlV1, GammaControlState>,
+    D: GammaControlHandler,
+    D: 'static,
+{
+    fn request(
+        state: &mut D,
+        _client: &Client,
+        _resource: &ZwlrGammaControlManagerV1,
+        request: <ZwlrGammaControlManagerV1 as Resource>::Request,
+        _data: &(),
+        _dhandle: &DisplayHandle,
+        data_init: &mut DataInit<'_, D>,
+    ) {
+        match request {
+            zwlr_gamma_control_manager_v1::Request::GetGammaControl { id, output } => {
+                let output_id = output.id();
+                let gamma_size = state.get_gamma_size(&output);
+                let previous_gamma_ramp = state.get_gamma(&output);
+                let mut got_gamma_map = true;
+                if previous_gamma_ramp.is_none() {
+                    got_gamma_map = false;
+                }
+
+                let zwlr_gamma_control = data_init.init(
+                    id,
+                    GammaControlState {
+                        gamma_size: gamma_size.clone(),
+                        previous_gamma_ramp,
+                        output,
+                    },
+                );
+                if !got_gamma_map {
+                    zwlr_gamma_control.failed();
+                    error!("Failed to get current gamma ramp");
+                    return;
+                }
+
+                if state
+                    .gamma_control_manager_state()
+                    .gamma_controls
+                    .contains_key(&output_id)
```

Comment:
What I'm afraid can happen is:
1. Client A sets gamma
2. Client B creates gamma control, gets `previous_gamma_ramp` equal to Client A's gamma, gets a fail here
3. Client A destroys global, resets gamma
4. Client B destroys global later, "resets" gamma to Client A's value instead of the original

So this `if` should also set `previous_gamma_ramp` to `None`.